vmngclient/api/templates.py

```diff
@@ -0,0 +1,299 @@
+import logging
+from typing import List, cast
+
+from ciscoconfparse import CiscoConfParse  # type: ignore
+from tenacity import retry, retry_if_result, stop_after_attempt, wait_fixed
+
+from vmngclient.dataclasses import DeviceInfo, Template
+from vmngclient.session import Session
+from vmngclient.utils.creation_tools import create_dataclass, get_logger_name
+from vmngclient.utils.device_model import DeviceModel
+from vmngclient.utils.operation_status import OperationStatus
+
+logger = logging.getLogger(get_logger_name(__name__))
+
+
+class TemplateNotFoundError(Exception):
+    """Used when a template item is not found."""
+
+    def __init__(self, template):
+        self.message = f"No such template: '{template}'"
+
+
+class TemplateExist(Exception):
+    """Used when a template item is exist."""
+
+    def __init__(self, name):
+        self.message = f"Template with that name '{name}' exists."
+
+
+class TemplateAttached(Exception):
+    """Used when delete attached template."""
+
+    def __init__(self, template):
+        self.message = f"Template: {template} is attached to device."
+
+
+class TemplateAPI:
+    """TemplateAPI."""
+
+    def __init__(self, session: Session) -> None:
+        self.session = session
+
+    @property
+    def templates(self) -> List[Template]:
+        """Templates.
+
+        Returns:
+            List[Template]: List of existing templates.
+        """
+        endpoint = "/dataservice/template/device"
+        data = self.session.get_data(endpoint)
+        return [create_dataclass(Template, template) for template in data]
+
+    def get(self, name: str) -> Template:
+        """Get template.
+
+        Args:
+            name (str): Name of template.
+
+        Raises:
+            TemplateNotFoundError: If template does not exist.
+
+        Returns:
+            Template: Selected template.
+        """
+        for template in self.templates:
+            if name == template.name:
+                return template
+        raise TemplateNotFoundError(name)
+
+    def get_id(self, name: str) -> str:
+        """Get id template.
+
+        Args:
+            name (str): Name of template to get id.
+
+        Returns:
+            str: Template id.
+        """
+        return self.get(name).id
+
+    def wait_complete(self, operation_id: str, timeout_seconds: int = 300, sleep_seconds: int = 5) -> bool:
```

Comment:
I know this function uses decorator @retry, but it seems that we already have more `wait_complete` or `wait_for_completed` similar functions in different modules.
Do you think it would be a good idea to make it also a decorator or more general function? I mean, instead of having special methods for each API that will evaluate completion, we could have one? 
Not mandatory for this particular PR, but would give it a thought.